File: src/main/java/com/github/eendroroy/sdk/openweather/interceptor/AbstractOWInterceptor.java
package com.github.eendroroy.sdk.openweather.interceptor;

import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;

import java.io.IOException;

/**
 * @author indrajit
 */

public abstract class AbstractOWInterceptor implements OWInterceptor {
    @Override
    public Response intercept(Chain chain) throws IOException {
        Request request = chain.request();
        request = interceptRequest(request);

        Response response = chain.proceed(request);
        response = interceptResponse(response);

        return response;
    }

    public abstract Request interceptRequest(Request request);

    public abstract Response interceptResponse(Response response);

    @Override
    public String rawResponseBody(Response response) {
        try {
            ResponseBody responseBody = response.body();
            if ( null == responseBody) {
                return "";
            }
            BufferedSource source = responseBody.source();
            source.request(Long.MAX_VALUE);
            Buffer buffer = source.getBuffer().clone();
            return buffer.readUtf8();
        } catch (IOException | IllegalStateException e) {
            return "<->";
        }
    }

    @Override
    public String rawRequestBody(final Request request) {
        try {
            final Request copy = request.newBuilder().build();
            final Buffer buffer = new Buffer();
            RequestBody body = copy.body();
            if (null == body) {
                return "> EMPTY_BODY <";
            }
            body.writeTo(buffer);
            return buffer.readUtf8();
        } catch (final IOException e) {
            return "<->";
        }
    }
}

File: src/main/kotlin/com/github/eendroroy/sdk/openweathermap/interceptor/AbstractOWInterceptor.kt
package com.github.eendroroy.sdk.openweathermap.interceptor

import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import okio.Buffer
import java.io.IOException

/**
 * @author indrajit
 */
abstract class AbstractOWInterceptor : OWInterceptor {
    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        var request = chain.request()
        request = interceptRequest(request)
        var response = chain.proceed(request)
        response = interceptResponse(response)
        return response
    }

    abstract fun interceptRequest(request: Request): Request
    abstract fun interceptResponse(response: Response): Response
    override fun rawResponseBody(response: Response): String {
        val responseBody = response.body() ?: return ""
        val source = responseBody.source()
        source.request(Long.MAX_VALUE)
        val buffer = source.buffer.clone()
        return buffer.readUtf8()
    }

    override fun rawRequestBody(request: Request): String {
        val copy = request.newBuilder().build()
        val buffer = Buffer()
        val body = copy.body() ?: return ""
        body.writeTo(buffer)
        return buffer.readUtf8()
    }
}